src/Compilers/Core/VBCSCompiler/AnalyzerConsistencyChecker.cs

```diff
@@ -0,0 +1,104 @@
+            }
+        }
+
+        private static bool CheckCore(string baseDirectory, IEnumerable<CommandLineAnalyzerReference> analyzerReferences, IAnalyzerAssemblyLoader loader, IEnumerable<string> referenceWhiteList)
+        {
+            var resolvedPaths = new List<string>();
+
+            foreach (var analyzerReference in analyzerReferences)
+            {
+                string resolvedPath = FileUtilities.ResolveRelativePath(analyzerReference.FilePath, basePath: null, baseDirectory: baseDirectory, searchPaths: SpecializedCollections.EmptyEnumerable<string>(), fileExists: File.Exists);
+                if (File.Exists(resolvedPath))
+                {
+                    resolvedPath = FileUtilities.TryNormalizeAbsolutePath(resolvedPath);
+                    if (resolvedPath != null)
+                    {
+                        resolvedPaths.Add(resolvedPath);
+                    }
+                }
+
+                // Don't worry about paths we can't resolve. The compiler will report an error for that later.
+            }
+
+            // First, check that the set of references is complete, modulo items in the whitelist.
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                var missingDependencies = AssemblyUtilities.IdentifyMissingDependencies(resolvedPath, resolvedPaths);
+
+                foreach (var missingDependency in missingDependencies)
+                {
+                    if (!referenceWhiteList.Any(name => missingDependency.Name.StartsWith(name)))
+                    {
+                        CompilerServerLogger.Log($"Analyzer assembly {resolvedPath} depends on '{missingDependency}' but it was not found.");
+                        return false;
+                    }
+                }
+            }
+
+            // Second, load all of the assemblies upfront.
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                loader.AddDependencyLocation(resolvedPath);
+            }
+
+            var loadedAssemblies = new List<Assembly>();
+            foreach (var resolvedPath in resolvedPaths)
+            {
+                loadedAssemblies.Add(loader.LoadFromPath(resolvedPath));
+            }
+
+            // Third, check that the MVIDs of the files on disk match the MVIDs of the loaded assemblies.
+            for (int i = 0; i < resolvedPaths.Count; i++)
+            {
+                var resolvedPath = resolvedPaths[i];
+                var loadedAssembly = loadedAssemblies[i];
+                var resolvedPathMvid = AssemblyUtilities.ReadMvid(resolvedPath);
+                var loadedAssemblyMvid = loadedAssembly.ManifestModule.ModuleVersionId;
+
+                if (resolvedPathMvid != loadedAssemblyMvid)
+                {
+                    CompilerServerLogger.Log($"Analyzer assembly {resolvedPath} has MVID '{resolvedPathMvid}' but loaded assembly '{loadedAssembly.FullName}' has MVID '{loadedAssemblyMvid}'.");
```

Comment:
I am wondering why we need to perform steps 2. and 3. here, i.e. loading assemblies eagerly and perform MVID checks. Can the fallback path of using csc/vbc succeed if we return false here due to mvid mismatches?